File: app/src/main/java/com/example/facturas/data/model/InvoiceVO.java
package com.example.facturas.data.model;

import static android.graphics.Color.BLACK;
import static android.graphics.Color.RED;

import android.content.res.Resources;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.room.Ignore;

import com.example.facturas.App;
import com.example.facturas.R;
import com.example.facturas.data.database.entity.InvoiceEntity;
import com.example.facturas.utils.MyConstants;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

public final class InvoiceVO implements Parcelable, Invoice {
    private int id;
    private String descEstado;
    private final Float importeOrdenacion;
    private final LocalDate fecha;
    private Integer textColor;

    @Ignore
    public InvoiceVO(String descEstado, Float importeOrdenacion, LocalDate fecha) {
        this(descEstado, importeOrdenacion, fecha, null);
    }

    public InvoiceVO(String descEstado, Float importeOrdenacion, LocalDate fecha, Integer textColor) {
        setDescEstado(descEstado);
        this.importeOrdenacion = importeOrdenacion;
        this.fecha = fecha;
        this.textColor = textColor;
    }

    public InvoiceEntity toInvoiceEntity() {
        return new InvoiceEntity(descEstado, importeOrdenacion, fecha, textColor);
    }

    // Constructor to read from Parcel
    protected InvoiceVO(Parcel in) {
        id = in.readInt();
        descEstado = in.readString();
        importeOrdenacion = in.readFloat();
        fecha = LocalDate.parse(in.readString());
        textColor = in.readInt();
    }

    @Override
    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    @Override
    public String getDescEstado() {
        return descEstado;
    }

    public void setDescEstado(String descEstado) {
        Resources context = App.getContext().getResources();
        this.textColor = BLACK;
        switch (descEstado) {
            case MyConstants.INVOICE_DESC_ESTADO_PAGADA:
                this.descEstado = context.getString(com.example.facturas.R.string.fragment_filter_cb_paid);
                break;
            case MyConstants.INVOICE_DESC_ESTADO_ANULADA:
                this.descEstado = context.getString(com.example.facturas.R.string.fragment_filter_cb_canceled);
                break;
            case MyConstants.INVOICE_DESC_ESTADO_CUOTA_FIJA:
                this.descEstado = context.getString(com.example.facturas.R.string.fragment_filter_cb_fixed_fee);
                break;
            case MyConstants.INVOICE_DESC_ESTADO_PENDIENTE:
                this.textColor = RED;
                this.descEstado = context.getString(com.example.facturas.R.string.fragment_filter_cb_pending_payment);
                break;
            case MyConstants.INVOICE_DESC_ESTADO_PLAN_PAGO:
                this.descEstado = context.getString(R.string.fragment_filter_cb_payment_plan);
                break;
            default:
                this.descEstado = descEstado;
        }
    }

    @Override
    public Float getImporteOrdenacion() {
        return importeOrdenacion;
    }

    public String getFecha(String format) {
        return fecha.format(DateTimeFormatter.ofPattern(format));
    }

    @Override
    public LocalDate getFecha() {
        return this.fecha;
    }

    @Override
    public Integer getTextColor() {
        return textColor;
    }

    public void setTextColor(int textColor) {
        this.textColor = textColor;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel out, int flags) {
        out.writeInt(id);
        out.writeString(descEstado);
        out.writeFloat(importeOrdenacion);
        out.writeString(fecha.toString());
        out.writeValue(textColor);
    }

    // Parcelable CREATOR
    public static final Parcelable.Creator<InvoiceVO> CREATOR = new Parcelable.Creator<InvoiceVO>() {
        @Override
        public InvoiceVO createFromParcel(Parcel in) {
            return new InvoiceVO(in);
        }

        @Override
        public InvoiceVO[] newArray(int size) {
            return new InvoiceVO[size];
        }
    };
}

File: app/src/main/java/com/example/facturas/data/model/InvoiceVO.kt
package com.example.facturas.data.model

import android.graphics.Color
import android.os.Parcel
import android.os.Parcelable
import android.os.Parcelable.Creator
import com.example.facturas.App
import com.example.facturas.R
import com.example.facturas.data.database.entity.InvoiceEntity
import com.example.facturas.utils.MyConstants
import java.time.LocalDate
import java.time.format.DateTimeFormatter

class InvoiceVO : Parcelable, Invoice {
    private var id = 0
    private var descEstado: String? = null
    private val importeOrdenacion: Float
    private val fecha: LocalDate
    private var textColor: Int?

    constructor(descEstado: String?, importeOrdenacion: Float, fecha: LocalDate, textColor: Int? = null) {
        setDescEstado(descEstado)
        this.importeOrdenacion = importeOrdenacion
        this.fecha = fecha
        this.textColor = textColor
    }

    fun toInvoiceEntity(): InvoiceEntity {
        return InvoiceEntity(descEstado, importeOrdenacion, fecha, textColor)
    }

    // Constructor to read from Parcel
    protected constructor(`in`: Parcel) {
        id = `in`.readInt()
        descEstado = `in`.readString()
        importeOrdenacion = `in`.readFloat()
        fecha = LocalDate.parse(`in`.readString())
        textColor = `in`.readInt()
    }

    override fun getId(): Int {
        return id
    }

    fun setId(id: Int) {
        this.id = id
    }

    override fun getDescEstado(): String {
        return descEstado!!
    }

    fun setDescEstado(descEstado: String?) {
        val context = App.getContext().resources
        textColor = Color.BLACK
        when (descEstado) {
            MyConstants.INVOICE_DESC_ESTADO_PAGADA -> this.descEstado = context.getString(
                R.string.fragment_filter_cb_paid
            )

            MyConstants.INVOICE_DESC_ESTADO_ANULADA -> this.descEstado =
                context.getString(R.string.fragment_filter_cb_canceled)

            MyConstants.INVOICE_DESC_ESTADO_CUOTA_FIJA -> this.descEstado =
                context.getString(R.string.fragment_filter_cb_fixed_fee)

            MyConstants.INVOICE_DESC_ESTADO_PENDIENTE -> {
                textColor = Color.RED
                this.descEstado = context.getString(R.string.fragment_filter_cb_pending_payment)
            }

            MyConstants.INVOICE_DESC_ESTADO_PLAN_PAGO -> this.descEstado =
                context.getString(R.string.fragment_filter_cb_payment_plan)

            else -> this.descEstado = descEstado
        }
    }

    override fun getImporteOrdenacion(): Float {
        return importeOrdenacion
    }

    fun getFecha(format: String?): String {
        return fecha.format(DateTimeFormatter.ofPattern(format))
    }

    override fun getFecha(): LocalDate {
        return fecha
    }

    override fun getTextColor(): Int? {
        return textColor
    }

    fun setTextColor(textColor: Int) {
        this.textColor = textColor
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(out: Parcel, flags: Int) {
        out.writeInt(id)
        out.writeString(descEstado)
        out.writeFloat(importeOrdenacion)
        out.writeString(fecha.toString())
        out.writeValue(textColor)
    }

    companion object CREATOR : Creator<InvoiceVO> {
        override fun createFromParcel(parcel: Parcel): InvoiceVO {
            return InvoiceVO(parcel)
        }

        override fun newArray(size: Int): Array<InvoiceVO?> {
            return arrayOfNulls(size)
        }
    }
}